### DISCUSSION TITLE
Difference between go-micro and micro repos

### DISCUSSION THREAD

**Discussion by cloudcompute** [micro/go-micro#2724] [Q&A] [answered]
There are two repositories by the name go-micro and micro. What is the difference between them and what shall we use for development of microservices?

Is it possible to add open-policy-agent as a plugin for authorization and connect to Kratos using OIDC for authentication?

**Comment by asim** [comment, marked as answer]
Go-micro is a framework like kratos and others. It's highly pluggable and sits at the center of you're code. You write your services with it and then choose the infrastructure you want to run beneath that. Everything is very modular. 

Micro on the other hand is a complete platform which includes it's own storage, service discovery, authentication, pubsub, API gateway, command line. It's meant more where you want to boot up an entire platform and dev environment without having to piece it all together. Micro was originally built around Go Micro but the projects have since diverged since the Go Micro users mostly wanted more control over their infrastructure choices. So Micro is more like an opinionated PaaS.

**Comment by cloudcompute** [comment]
@asim

Thanks for the quick response.

Does this mean: go-micro + plugins (repo) = micro.

Where does the 'plugins' repo actually fit in, I am not able to understand. Cannot we use it with micro?

**Reply by asim** [reply]
So the go micro packages are based on go interfaces and the plugins. Let you swap out the underlying implementation of those plugins. So if you want to use gnats or Kafka for pub sub messaging through the broker then you can use them in the plugins. But that alone does not equal micro. Micro itself is more like a set of apis. 

And at this moment you cannot use the plugins with micro. So one thing I'm thinking about is unifying the interfaces again so that this would be possible.

**Comment by cloudcompute** [comment]
> So one thing I'm thinking about is unifying the interfaces again so that this would be possible.

I guess it'd be a time-consuming task.

> You write your services with it and then choose the infrastructure you want to run beneath that.

Is there any application that demonstrates couple of services making use of the pluggable infrastructure?

So if we decide to use go-micro, we have to take care of several things ourselves like Service discovery, authentication, API gateway, and what else?

**Comment by asim** [comment]
I think the plugins repo shows how you make use of it. There might be some things in https://github.com/go-micro which was community maintained 

If you do use go-micro then yes you're constructing the whole platform around it
